packages/core/data-transfer/src/strapi/providers/remote-source/index.ts

```diff
@@ -151,10 +157,29 @@ class RemoteStrapiSourceProvider implements ISourceProvider {
             const chunk = Buffer.from(rawBuffer.data);
 
             await this.writeAsync(assets[item.assetID].stream, chunk);
+            assetStatus[item.assetID] -= 1;
           }
 
           // The asset has been transferred
           else if (action === 'end') {
+            let endStreamResolve;
+            let endStreamPromise = new Promise<void>((resolve) => {
+              endStreamResolve = resolve;
+            });
+            if (assetStatus[item.assetID] === 0) {
+              endStreamResolve();
+            } else { // wait if end received before previous data finished writing
+              let fn = () => {
+                if (assetStatus[item.assetID] === 0) {
+                  endStreamResolve();
+                } else {
+                  setTimeout(fn, 300)
+                }
+              }
+              setTimeout(fn, 300);
+            }
+            await endStreamPromise;
+
```

Comment:
```suggestion
            /**
             * NOTE: This suggests we're processing streaming events wrong.
             * We need to investigate and delay 'end' from being called
             * until streaming has finished, rather than waiting here
             * */
            // Wait for all chunks to finish streaming before continuing to the end
            await new Promise<void>((resolve) => {
              const interval = setInterval(() => {
                if (assetStatus[item.assetID] === 0) {
                  clearInterval(interval);
                  resolve();
                }
              }, 100);
            });

```
As far as I can tell this is functionally equivalent and would be easier to follow. Can you check if it still works for you?